fix: dropdown menu item

samyarkd opened the pull request:
Disable dropdown menu item when type is equal to danger.

<!--
First of all, thank you for your contribution! 😄

For requesting to pull a new feature or bugfix, please send it from a feature/bugfix branch based on the `master` branch.

Before submitting your pull request, please make sure the checklist below is confirmed.

Your pull requests will be merged after one of the collaborators approve.

Thank you!

-->

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [X] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [X] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
<!--
1. Put the related issue or discussion links here.
-->

### 💡 Background and solution
**Problem**:  When we set the dropdown menu item properties `danger` and `disabled` to `true`, the disabled styling doesn't apply. As shown in below code sandbox 👇
https://codesandbox.io/embed/antd-reproduction-template-forked-ip8ypb?fontsize=14&hidenavigation=1&theme=dark

**Solution**: check if dropdown menu item is **disabled** or not
If yes  -> we are not going to apply danger staying, same as how button handles it (shown in code sandbox).  

<!--
1. Describe the problem and the scenario.
2. GIF or snapshot should be provided if includes UI/interactive modification.
3. How to fix the problem, and list the final API implementation and usage sample if that is a new feature.
-->

### 📝 Changelog

<!--
Describe changes from the user side, and list all potential break changes or other risks.
--->
This file "components/menu/MenuItem.tsx" is updated.

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   fix: apply disabled style to dropdown menu item when type is danger : https://github.com/ant-design/ant-design/issues/39322  |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

samyarkd committed:
fix: dropdown menu item in danger 

Disable dropdown menu item when type is equal to danger

Wxh16144 commented on components/menu/MenuItem.tsx:
```diff
@@ -66,7 +66,7 @@ export default class MenuItem extends React.Component<MenuItemProps> {
           {
-            [`${prefixCls}-item-danger`]: danger,
+            [`${prefixCls}-item-danger`]: this.props.disabled ? false : danger,
             [`${prefixCls}-item-only-child`]: (icon ? childrenLength + 1 : childrenLength) === 1,
```
I think we can solve the problem here by increasing the css selector weights.

https://github.com/ant-design/ant-design/blob/f4d011623ebcb0eafe5e3ac9228d3f7b681c5152/components/dropdown/style/index.less#L224-L237

what do you think?

```diff
-      &-disabled {
+      &&-disabled {
```

samyarkd replied:
Thank you 

I tried this out in version 5x, but it didn't work out.
It will break the disable option as shown in below (Image-1). But yes this is much cleaner and better i guess but i couldn't make it i tried a few more ways of increasing the **CSS specificity**, but they didn't work.

![Image-1](https://user-images.githubusercontent.com/75088294/206263342-74265eb0-3766-4d8a-8ee0-cdb651c1b5a6.png)

![code-1](https://user-images.githubusercontent.com/75088294/206263468-5e6a98e3-4835-4d49-aa27-18e331e6209a.png)

samyarkd replied:
What about adding `!important` statement to `color` and `background-color` ?

Wxh16144 replied:
The use of `!important ` is bad, in 5.x it can be changed to this

```diff
 
-            '&-disabled': {
+            [`&${menuCls}-item-disabled`]: {
               color: colorTextDisabled,

```

and switch the order of returns

```diff
     return [
+      genStatusStyle(dropdownToken),
       genBaseStyle(dropdownToken),
       genButtonStyle(dropdownToken),
-      genStatusStyle(dropdownToken),
     ];
```

Wxh16144 replied:
this is just my suggestion, still need official help to review @afc163

afc163 replied:
@Wxh16144's suggestion is what I think too.

yoyo837 commented:
Fix for v4? Please send PR to branch `v4-stable`

samyarkd committed:
Merge branch 'ant-design:master' into bugfix

samyarkd committed:
increase disabled CSS specificity

samyarkd committed:
remove unnecessary changes

Wxh16144 commented on components/dropdown/style/index.tsx:
```diff
@@ -456,9 +456,9 @@ export default genComponentStyleHook(
       dropdownEdgeChildPadding: paddingXXS,
     });
     return [
+      genStatusStyle(dropdownToken),
```
> I prefer to remain the order because `genBaseStyle` should be at first. Maybe problem is solved without changing order since the priority has been raised?

I've also tested this and it still doesn't fix the problem without reordering

samyarkd replied:
I made some changes without reordering the functions

samyarkd committed:
Merge branch 'ant-design:master' into bugfix

samyarkd committed:
fix: remove reordering functions 

Remove reordering functions by increasing `disable` class on line 350 CSS specificity

MadCcc commented on components/dropdown/style/index.tsx:
```diff
@@ -347,7 +347,7 @@ const genBaseStyle: GenerateStyle<DropdownToken> = (token) => {
             },
 
-            '&-disabled': {
+            [`&${menuCls}-item-disabled${menuCls}-item-disabled`]: {
```
```suggestion
            [`&-disabled, &-disabled${menuCls}-item-danger`]: {
```
How about this?

samyarkd replied:
I tested it and it didn't work

afc163 commented:
Fixed in https://github.com/ant-design/ant-design/pull/39434

Wxh16144 commented:
> Fixed in #39434

#39434 修复的是 4.x 的，这个 PR 修复的是 5.x

yoyo837 commented:
@Wxh16144 你改过V4, 把把关

Wxh16144 commented:
> @Wxh16144 你改过V4, 把把关

其实改变顺序问题不大，在 v4 也是吧顺序调到了disabled 前面，我 [前面讨论](https://github.com/ant-design/ant-design/pull/39368#discussion_r1043120663) 的时候建议就是如此。 

我再看看有没有更好的方式～

afc163 commented:
@Wxh16144 进展如何

Wxh16144 commented:
> 进展如何

我认为还是最开始的解决方案 OK 一些，要么就在 rc 组件将 disable 的 class移动到最后面 

ref: https://github.com/react-component/menu/blob/27271d6e963c7a48947de74efcb46d0cf5b6b869/src/MenuItem.tsx#L213
